File: solutions/erlang/anagram/1/src/anagram.erl
-module(anagram).

-export([find_anagrams/2]).


find_anagrams(Subject, Candidates) ->
  UniqueCandidates = lists:filter(
    fun(Candidate) ->
      string:to_upper(Candidate) =/= string:to_upper(Subject)
    end,
    Candidates),
  lists:filter(
    fun(Candidate) ->
      graph(Candidate) == graph(Subject)
    end,
    UniqueCandidates).

graph(Word) ->
  UpperCaseWord = string:to_upper(Word),
  Graphemes = string:to_graphemes(UpperCaseWord),
  lists:sort(Graphemes).

File: solutions/java/anagram/1/src/main/java/Anagram.java
import java.util.Arrays;
import java.util.List;
import java.util.Locale;

class Anagram {
    private final String normalizedWord;
    private final String signature;

    public Anagram(String word) {
        this.normalizedWord = normalizeWord(word);
        this.signature = sortLetters(normalizedWord);
    }

    public List<String> match(List<String> candidates) {
        return candidates.stream()
                         .filter(this::isAnagramCandidate)
                         .toList();
    }

    private boolean isAnagramCandidate(final String candidate) {
        String normalizedCandidate = this.normalizeWord(candidate);
        return !normalizedCandidate.equals(normalizedWord)
               && this.sortLetters(normalizedCandidate).equals(signature);
    }

    private String normalizeWord(String word) {
        return word.toLowerCase(Locale.ROOT);
    }

    private String sortLetters(String word) {
        char[] chars = word.toCharArray();
        Arrays.sort(chars);
        return new String(chars);
    }
}